### PR TITLE
[WIP] Implement Elastic Search Code Indexer

### PR DESCRIPTION
**Pull request by jeffliu27**
Creates Issue indexer as a separate package: "codes" (will be changed to code later). 
Added elastic search support for indexer instead of "bleve"

Ref issue: #6648

### CODE REVIEW

File: modules/indexer/codes/bleve.go [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+	docMapping.AddFieldMappingsAt("Content", textFieldMapping)
+
+	mapping := bleve.NewIndexMapping()
+	if err = addUnicodeNormalizeTokenFilter(mapping); err != nil {
+		return nil, err
+	} else if err = mapping.AddCustomAnalyzer(repoIndexerAnalyzer, map[string]interface{}{
+		"type":          custom.Name,
+		"char_filters":  []string{},
+		"tokenizer":     unicode.Name,
+		"token_filters": []string{unicodeNormalizeName, camelcase.Name, lowercase.Name, unique.Name},
```
Comment (guillep2k) [inline modules/indexer/codes/bleve.go:89]: You may want to reconsider using camelcase and unique. See #7733, #7825 and #7826.

File: modules/indexer/codes/indexer.go [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+		return err
+	} else if !base.IsTextFile(fileContents) {
+		return nil
+	}
+	return batch.Index(
+		filenameIndexerID(repo.ID, update.Filename),
+		&IndexerData{
+			RepoID:   repo.ID,
+			Filepath: update.Filename,
+			Content:  string(fileContents),
```
Comment (guillep2k) [inline modules/indexer/codes/indexer.go:242]: You may want to make sure the file contents is valid utf-8. See #7809 and #7814.

**Review by guillep2k — COMMENTED** [review]
It looks like I've been PR'ing on the same subject without realizing you were already working on it. For that I apologize. I leave you here a couple of highlights about my findings.

**Comment by guillep2k** [comment]
Is this PR still alive? Otherwise I'd like to take #3670.

**Comment by jeffliu27** [comment]
> Is this PR still alive? Otherwise I'd like to take #3670.

Yes, planning on finishing it. (Haven't gotten a chance to work on it recently)

**Comment by Cellebyte** [comment]
@jeffliu27 do you had some time for this feature?.

**Comment by Cellebyte** [comment]
> @jeffliu27 do you had some time for this feature?.

@jeffliu27 are you still working on this?

**Comment by Cellebyte** [comment]
@lunny can this here be implemented with the same logic as you are implementingthe issue indexer?

**Comment by lunny** [comment]
@Cellebyte Yes. I will do that and replace this one if my previous issue indexer merged.

**Comment by ksa-real** [comment]
@jeffliu27 Any news on your diff? It seems your code has conflicts.

**Pull request closed without merging**